### ISSUE TITLE
No difference beetween `cache-and-network` and `network-only` strategies!

### ISSUE THREAD

**Issue by foad-salawati** [apollographql/apollo-client#7605]
I think the `cache-and-network` strategy doesn't work properly and acts similar to `network-only`.

**What To Expect**

1. apollo sends two queries to cache and network in parallel.
2. if doesn't exist in cache waiting for network and store in cache for next query.
3. if exists, render data and wait for network in background for syncing data.

but in step 3 :

apollo doesn't render cache data until network data is received. **This is not fast** and behaves like  `network-only` strategy.

**Comment by wminshew** [comment]
have noticed this as well; seems like there should be another exposed boolean like `refetching` or something so `loading` isn't used when there is relevant data in the cache and the network request is running in the background

**Comment by fforres** [comment]
This is the solution we are using to expose a `refetching` prop so we can control rendering behaviour better @wminshew @foad-salawati 

```tsx
import type { OperationVariables, QueryResult } from "@apollo/react-common";
import type { QueryHookOptions } from "@apollo/react-hooks";
import { useQuery as useApolloQuery } from "@apollo/react-hooks";
import type { DocumentNode } from "graphql";

export function useQuery<TData = any, TVariables = OperationVariables>(
  query: DocumentNode,
  options?: QueryHookOptions<TData, TVariables>,
): QueryResult<TData, TVariables> & {
  refetching: boolean;
} {
  const response = useApolloQuery<TData, TVariables>(query, options);
  return {
    ...response,
    refetching: response.networkStatus === 4,
  };
}
```

**Comment by lassesteffen** [comment]
Also have this problem, `cache-and-network` is not loading from cache first

**Comment by lassesteffen** [comment]
@foad-salawati did you find a workaround?

**Comment by foad-salawati** [comment]
> @foad-salawati did you find a workaround?

Not yet! follow this opened issue [#1781](https://github.com/redwoodjs/redwood/issues/1781) for some solutions or a new released in the future.

**Comment by lassesteffen** [comment]
I just saw that if you use `notifyOnNetworkStatusChange: true` it works for me and does return the cached data on the first return

**Comment by HoseinGhanbari** [comment]
> I just saw that if you use `notifyOnNetworkStatusChange: true` it works for me and does return the cached data on the first return

using both `notifyOnNetworkStatusChange: true` and `fetchPolicy: 'cache-and-network',` didn't solve the issue

**Comment by llamington** [comment]
> using both notifyOnNetworkStatusChange: true and fetchPolicy: 'cache-and-network', didn't solve the issue

Neither with my setup:
`"@apollo/client": "^3.10.4"`

**Comment by mjfs00** [comment]
> using both `notifyOnNetworkStatusChange: true` and `fetchPolicy: 'cache-and-network',` didn't solve the issue

same here. didn't work with my setup too

```
"@apollo/client": "^3.10.4",
"apollo-angular": "^7.0.1",
```

**Comment by HoseinGhanbari** [comment]
> using both `notifyOnNetworkStatusChange: true` and `fetchPolicy: 'cache-and-network',` didn't solve the issue

same here. didn't work with my setup too:

```
"@apollo/client": "^3.10.4",
"apollo-angular": "^7.0.1",
```

**Comment by CarlosNZ** [comment]
This was really annoying me too, but I think I finally have a solution.

The problem is that that, even though `data` gets the cached value immediately, the `loading` state still gets set to `true` while the additional network request is running. And so if you're rendering a "Loading" component based on `loading`, then you won't see the new data.

The solution is to check both the `data` *and* the `loading` state before rendering your "Loading" UI.

```js
const {data, loading, error } = useQuery ( {variables, fetchPolicy: 'cache-and-network' });

if (loading && !data) return <LoadingComponent />

return <DataComponent data={data} />
```

This is working for me -- I get the cached result immediately when there is one, and a Loader when there isn't any cache available (and the UI gets updated when *new* data comes in from the network request, as expected).

**Comment by HoseinGhanbari** [comment]
Found a temporary solution which is simulating `cache-and-network` somehow, just set `fetchPolicy: 'cache-first',` and also do `ref.refetch();` whenever a "query" gets send to the server. Actually this is not the way I like but the REAL `cache-and-network` is not working.

```ts
    query<T = any, V extends Record<string, any> = {}>(query: DocumentNode, variables?: V, fetchPolicy: WatchQueryFetchPolicy = 'cache-and-network'): GeneralQueryType<T, V> {
        const ref = this.apollo.watchQuery<T, V>({
            query,
            variables,
            context: this.context,
            fetchPolicy: 'cache-first',
            refetchWritePolicy: 'overwrite',
        });

        const obs = ref.valueChanges.pipe(
            filter(result => result.networkStatus === NetworkStatus.ready),
            map(response => response.data));

        // NOTICE :: temporary workaround
        ref.refetch();

        return {
            obs,
            ref
        };
    }
```

**Comment by mink-ang** [comment]
Turns out apollo won't be able to cache previous results correctly depending on your query and whether it has field called `id`, you will need to set type policies so the query results can be properly normalised and saved in cache. 

https://www.apollographql.com/docs/react/caching/cache-configuration/

Hopefully this helps someone out there.

**Comment by aike19115** [comment]
The option `fetchPolicy: 'cache-only'` still doesn't behave as expected. By enabling `notifyOnNetworkStatusChange: true` it does. It's been almost four years that this issue was raised, hopefully this comment raises aware. My setup is:
```
"@apollo/client": "^3.11.10",
"@react-native-async-storage/async-storage": "1.23.1",
"apollo3-cache-persist": "^0.15.0",
"expo": "~51.0.28",
"graphql": "^16.9.0",
"react": "18.2.0",
"react-native": "0.74.5",
```

**Comment by aike19115** [comment]
> This was really annoying me too, but I think I finally have a solution.
> 
> The problem is that that, even though `data` gets the cached value immediately, the `loading` state still gets set to `true` while the additional network request is running. And so if you're rendering a "Loading" component based on `loading`, then you won't see the new data.
> 
> The solution is to check both the `data` _and_ the `loading` state before rendering your "Loading" UI.
> 
> ```js
> const {data, loading, error } = useQuery ( {variables, fetchPolicy: 'cache-and-network' });
> 
> if (loading && !data) return <LoadingComponent />
> 
> return <DataComponent data={data} />
> ```
> 
> This is working for me -- I get the cached result immediately when there is one, and a Loader when there isn't any cache available (and the UI gets updated when _new_ data comes in from the network request, as expected).

Was running into the same issue and this explains it 👌

**Comment by phryneas** [comment]
> This was really annoying me too, but I think I finally have a solution.
> 
> The problem is that that, even though `data` gets the cached value immediately, the `loading` state still gets set to `true` while the additional network request is running. And so if you're rendering a "Loading" component based on `loading`, then you won't see the new data.
> 
> The solution is to check both the `data` _and_ the `loading` state before rendering your "Loading" UI.
> 
> ```js
> const {data, loading, error } = useQuery ( {variables, fetchPolicy: 'cache-and-network' });
> 
> if (loading && !data) return <LoadingComponent />
> 
> return <DataComponent data={data} />
> ```
> 
> This is working for me -- I get the cached result immediately when there is one, and a Loader when there isn't any cache available (and the UI gets updated when _new_ data comes in from the network request, as expected).

To clarify, what is described here is the intended behaviour - you will be `loading`, but have initial data available.

I'll be honest, the comments in this thread are a bit unclear to me - do you all have this behaviour and didn't expect it, or is there some bug beyond that where you didn't get `data` initially even though it was present in the cache?

**Comment by CarlosNZ** [comment]
> To clarify, what is described here is the intended behaviour - you will be `loading`, but have initial data available.
> 
> I'll be honest, the comments in this thread are a bit unclear to me - do you all have this behaviour and didn't expect it, or is there some bug beyond that where you didn't get `data` initially even though it was present in the cache?

Yes, it does make sense that this is the intended behaviour, it's just not immediately obvious.
